Add documentation (#4)
yozhgoor/CreateProcessW: README.md

# CreateProcessW

This crate provides an API similar to [`std::process`][std-process] to create
and handle processes on Windows using the Win32 API through the
[windows-rs][windows-rs] crate (see [this example][create-processes-example]).

Its main difference with `std::process::Command` is that it allows running
a command string instead of having to pass the command executable and the
arguments separately.

This is equivalent of running:

```rust
std::process::Command::new("cmd.exe")
    .arg("/c")
    .arg("any_command_string")
    .spawn().expect("cannot spawn command");
```

The only difference will be that the `Child` instance will use the PID of
the command instead of the PID of `cmd.exe`. This is important because
calling `.kill()` in the code above does not work as it kills the PID
of `cmd.exe` instead of the actual command that has been ran.

## Usage

Add the following to your `Cargo.toml`:

```toml
[dependencies]
CreateProcessW = "0.1.0"
```

> This crate doesn't follow Rust's naming recommendations and disable warnings
> about `non_snake_case`. If you want to stay consistent with other imported
> crates, use the following:

Reviewer: I think it could sounds a bit like it disables non snake case for others which sounds scary. Let's not hurry anyone.

```suggestion
This crate doesn't follow Rust's naming recommendations.
If you want to stay consistent with other imported crates, use the following:
```

```toml
[dependencies]
create_process_w = { version = "0.1.0", package = "CreateProcessW" }
```

## Create a command

The [`Command`] struct is used to configure and spawn processes:

```rust
use CreateProcessW::Command;

let command = Command::new("cargo.exe clippy -- -D warnings")

>     .inherit_handles(false)

Reviewer: This is the default now so... just remove it to not make it confusing
```suggestion
```

# CreateProcessW

> This crate provide an API similar to `std::process` to create and handle process
> using the [windows-rs][windows-rs] crate and the Windows API (see [this example][creating-processes]).

Reviewer: ```suggestion
This crate provide an API similar to `std::process` to create and handle processes
on Windows using the Win32 API through the [windows-rs][windows-rs] crate
(see [this example][creating-processes]).

It's main difference with `std::process::Command` is that it allows running a command
string instead of having to pass the command executable and the arguments separately.

This is the equivalent of running:

std::process::Command("cmd.exe")
    .arg("/c")
    .arg(any_command_string)
    .spawn()

The only difference will be that the `Child` instance will use the PID of the command instead of
the PID of `cmd.exe`. This is important because calling `.terminate()` in the code above does
not work as it kills the PID of `cmd.exe` instead of the actual command that has been ran.
```

> # Usage

Reviewer: Can't be at same level than document's title

Add the following to your `Cargo.toml`:

```toml
[dependencies]
CreateProcessW = "0.1.0"
```

This crate doesn't follow Rust's naming convention, if you want to avoid the
`non_snake_case` name, you can use the following:

```toml
[dependencies]
create_process_w = { version = "0.1.0", package = "CreateProcessW" }

> ```

Reviewer: ~~Only provide the solution with the rename!! XD nobody wants to deal with the non_snake_case~~ [edit: my bad. People don't need to add `non_snake_case` in their own crate. It only needs to be on `CreateProcessW`]

# Create a new Command

To configure and create a new process, you can use the `Command` struct. The
only argument needed is the command line you want to execute.

```rust

> use CreateProcess::Command;

Reviewer: The crate's name is CreateProcessW, not CreateProccess :confounded: 

But anyway, use the version that respect the style guide lines: create_process_w

> let command = Command::new("ls");

Reviewer: `ls`??? on Windows? :rofl: at least provide a working example

Author: Not problem, in this case i receive the files of my crate as the output of `cargo-run`.
I followed the `std::process` documentation, and it work if i give a `current_dir` too.
`dir` is a better example maybe ?

```

## Customization

You can customize the command with some methods:

> * `inherit_handles` - If this parameter is TRUE, each inheritable handle in the
>     calling process is inherited by the new process. If the parameter is FALSE,
>     the handles are not inherited. Note that inherited handles have the same
>     value and access rights as the original handles. The default value is
>     `true`.

Reviewer: ```suggestion
* `inherit_handles` - If this parameter is `true`, each inheritable handle in the
    calling process is inherited by the new process. If the parameter is FALSE,
    the handles are not inherited. Note that inherited handles have the same
    value and access rights as the original handles. The default value is
    `true`.
```

* `current_dir` - The full path to the current directory for the process. If you
    don't provide a value, the new process will have the same current drive and
    directory as the calling process.

Example:

```rust
use CreateProcessW::Command;

let command = Command::new("ls")
    .inherit_handles(false)
    .current_dir("C:\\Users\\user\\repos");

> ```

Reviewer: Start your section with the example to show off the builder pattern first.

# Execute the command

> This library give you two way to execute a command, `spawn` return a handle to
> the child process and `status` wait for it to finish and collect its status.

Reviewer: ```suggestion
This library give you two ways to execute a command that match `std::process`'s API:
- `spawn` - return a handle to
the child process.
- `status` - wait for it to finish and collect its status.

`output` is not available at the moment.
```

## Spawning a process

The `spawn` method start a process and return a [`Child`] struct that allows to
handle the process. `Child` provide several methods to do this:

```rust
use CreateProcessW::Command;

let child = Command::new("notepad.exe").spawn().expect("cannot spawn notepad");

if let Some(status) = child.try_wait().expect("waiting process failed") {
    println!("Process exited with status code {}", status.status());
} else {
    println!("Process is running");
}

child.kill().expect("cannot kill process");
child.wait().expect("cannot wait process");
```

The `try_wait` method try to get the `ExitStatus` of the process. If the process
is running, this method return `None`, if the process is terminated this method
return `Some(ExitStatus)`.

The `kill` method terminate the process.

The `wait` method wait the end of the process and return an `ExitStatus`, it

> also clean up the process by closing the handles.

Reviewer: Feels like way too much details in my opinion.

The point of replicating the std's API is to not have to re-explain what people already know. Only explain the differences with the std's API (like: output is missing, inherit_handles inherit ALL handles, and most importantly `new`'s argument is for the entire command and not just the executable!)

# CreateProcessW

> This crate provide an API similar to [`std::process`][std-process] to create

Reviewer: ```suggestion
This crate provides an API similar to [`std::process`][std-process] to create
```

and handle processes on Windows using the Win32 API through the
[windows-rs][windows-rs] crate (see [this example][create-processes-example]).

> It's main difference with `std::process::Command` is that it allows running

Reviewer: ```suggestion
Its main difference with `std::process::Command` is that it allows running
```

a command string instead of having to pass the command executable and the
arguments separately.

This is equivalent of running:

```rust
std::process::Command("cmd.exe")
    .arg("/c")
    .arg(any_command_string)
    .spawn()
```

The only difference will be that the `Child` instance will use the PID of
the command instead of the PID of `cmd.exe`. This is important because
calling `.kill()` in the code above does not work as it kills the PID
of `cmd.exe` instead of the actual command that has been ran.

## Usage

Add the following to your `Cargo.toml`:

```toml
[dependencies]
create_process_w = { version = "0.1.0", package = "CreateProcessW" }
```

You can also use `CreateProcessW` directly, but this doesn't respect Rust's
naming recommendations.

## Create a command

The [`Command`] struct is used to configure and spawn processes:

```rust
use create_process_w::Command;

let command = Command::new("cargo.exe check")
    .inherit_handle(false)
    .current_directory(r"C:\Users\<user>\repos\<repo_name>");
```

### Spawning a process

> The [`spawn`][Command::spawn] function spawn the process and return a [`Child`] that

Reviewer: ```suggestion
The [`spawn`][Command::spawn] function spawns the process and returns a [`Child`] that
```

represents the spawned child process.

```rust
use create_process_w::Command;

let child = Command::new("notepad.exe")
    .spawn()
    .expect("notepad failed to start");

std::thread::Duration(std::time::Duration::from_secs(2));

child.kill().expect("cannot kill process");
let status = child.wait().expect("cannot wait process");

if status.success() {
    println!("Success!");
} else {
    println!("Process exited with status {}", status.code());
}
```

> The [`status`][Command::status] function spawn a child process, wait for it to finish and

Reviewer: ```suggestion
The [`status`][Command::status] function spawns a child process, waits for it to finish and
```

> return its [`ExitStatus`].

Reviewer: ```suggestion
returns its [`ExitStatus`].
```
